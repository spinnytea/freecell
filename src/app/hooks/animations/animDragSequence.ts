import { MutableRefObject } from 'react';
import { gsap } from 'gsap/all';
import { BOTTOM_OF_CASCADE } from '@/app/components/cards/constants';
import { calcCardId } from '@/app/game/card/card';

// FIXME keep refining tale
export function animDragSequence({
	timeline,
	// FIXME we need first card coords, not the position of the pointer/cursor/mouse
	pointerCoords: { x, y },
	list,
	offsetTop,
	gameBoardIdRef,
}: {
	timeline: gsap.core.Timeline;
	list: string[];
	pointerCoords: { x: number; y: number };
	offsetTop: number;
	gameBoardIdRef?: MutableRefObject<string>;
}) {
	const positions = list.map((shorthand) => {
		const el = document.getElementById(calcCardId(shorthand, gameBoardIdRef?.current));
		if (el) {
			const bounds = el.getBoundingClientRect();
			return {
				top: bounds.top,
				left: bounds.left,
			};
		}
		return {
			top: y,
			left: x,
		};
	});

	list.forEach((shorthand, index) => {
		const cardId = '#' + calcCardId(shorthand, gameBoardIdRef?.current);
		const zIndex = BOTTOM_OF_CASCADE + index;
		gsap.set(cardId, { zIndex });
		if (index > 0) {
			// follow the card this is stacked on top of
			// an attempt at follow-the-leader style drag animation, each card lags behind the previous
			const previousCardPosition = positions[index - 1];
			const top = previousCardPosition.top + offsetTop;
			const left = previousCardPosition.left;
			const duration = 2 + index / 10; // REVIEW (drag-and-drop) timing
			timeline.to(cardId, { top, left, duration, ease: 'linear' }, '<0');
		}
	});
}
