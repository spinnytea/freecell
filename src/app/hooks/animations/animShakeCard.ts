import { MutableRefObject } from 'react';
import { gsap } from 'gsap/all';
import { INVALID_SHAKE_MAGNITUDE, INVALID_SHAKE_PORTION } from '@/app/animation_constants';

export function animShakeCard({
	timeline,
	list,
	gameBoardIdRef,
}: {
	timeline: gsap.core.Timeline;
	list: string[];
	gameBoardIdRef?: MutableRefObject<string>;
}) {
	// sometimes we start left, sometimes we start right
	// yoyo is opposite
	const start: boolean = gsap.utils.random([true, false]);
	const signStart = start ? '-' : '+';
	const signYoyo = start ? '+' : '-';

	list.forEach((shorthand, index) => {
		const cardId = '#c' + shorthand + (gameBoardIdRef?.current ? '-' + gameBoardIdRef.current : '');
		const tl = gsap.timeline();
		// offset left
		tl.to(cardId, {
			x: `${signStart}=${INVALID_SHAKE_MAGNITUDE.toFixed(3)}`,
			duration: 0.1,
			ease: 'sine.in',
		});
		// yoyo
		tl.to(cardId, {
			x: `${signYoyo}=${(INVALID_SHAKE_MAGNITUDE * 2).toFixed(3)}`,
			duration: INVALID_SHAKE_PORTION,
			ease: 'sine.inOut',
			yoyo: true,
			repeat: 2,
		});
		// back to center
		tl.to(cardId, { x: '0', duration: INVALID_SHAKE_PORTION / 2, ease: 'sine.out' });

		timeline.add(tl, index === 0 ? `>0` : `<+${INVALID_SHAKE_PORTION.toFixed(3)}`);
	});
}
