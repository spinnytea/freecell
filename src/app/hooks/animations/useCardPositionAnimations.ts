import { MutableRefObject, useRef } from 'react';
import { useGSAP } from '@gsap/react';
import { gsap } from 'gsap/all';
import { MULTI_ANIMATION_TIMESCALE } from '@/app/animation_constants';
import { calcCardId } from '@/app/game/card/card';
import { animShakeCard } from '@/app/hooks/animations/animShakeCard';
import { animUpdatedCardPositions } from '@/app/hooks/animations/animUpdatedCardPositions';
import { calcUpdatedCardPositions } from '@/app/hooks/animations/calcUpdatedCardPositions';
import { useFixtureSizes } from '@/app/hooks/contexts/FixtureSizes/useFixtureSizes';
import { useGame } from '@/app/hooks/contexts/Game/useGame';

// IDEA (settings) setting for "reduced motion" - disable most animations
// IDEA (animation) faster "select-to-peek" animation - when the cards are shifting to peek the selected card
export function useCardPositionAnimations(gameBoardIdRef?: MutableRefObject<string>) {
	const { cards, selection, previousAction } = useGame();
	const fixtureSizes = useFixtureSizes();

	/**
		keep track of card positions, we need to animate anything that moves
		if it hasn't moved since last time, then we don't need to animate it

		IDEA (animation) (drag-and-drop) (techdebt) Store previous positions on DOM? Data attr? Context? GSAP (like, hault an animation, and continue from current position)?
		 - It's just t/l
		 - accessor method for unit testing
		---
		 - a local previousTLs doesn't play with other animations
		 - drag-and-drop needs to know card positions too


		IDEA (animation) (techdebt) Initial positions
		 - Animations don't have positions until they change
		 - So it animates everything all at once (fine)
		 - Iff we don't have previousTLs, then g.undo() and seed it with those.

		BUG (animation) (techdebt) Refresh and then immediately "new game" doesn't animate correctly (possibly because there are no saved card positions l yet)
	*/
	const previousTLs = useRef(new Map<string, number[]>());

	/**
		if we change the size of the screen, then everything will animate
		don't do any offsets, just move/update all the cards immediately

		positions of cards are controlled entirely by the animates

		TODO (techdebt) change to "usePrevious" which can/should return hasChanged
	*/
	const prevFixtureSizes = useRef(fixtureSizes);

	/**
		handle on the currently running / previous animation
		if we have one running when we start a new one,
		then finish it and kill it immediately
		this is a user-interactive game, and we want to prioritize responsiveness
	*/
	const previousTimeline = useRef<gsap.core.Timeline | null>(null);

	useGSAP(
		() => {
			const timeline = gsap.timeline({
				onComplete: () => {
					previousTimeline.current = null;
				},
			});

			const {
				updateCardPositions,
				updateCardPositionsPrev,
				secondMustComeAfter,
				unmovedCards,
				invalidMoveCards,
			} = calcUpdatedCardPositions({
				fixtureSizes,
				previousTLs: previousTLs.current,
				cards,
				selection,
				previousAction,
			});

			if (updateCardPositions.length) {
				if (previousTimeline.current && previousTimeline.current !== timeline) {
					// FIXME test
					previousTimeline.current.timeScale(MULTI_ANIMATION_TIMESCALE); // speed up the previous animations
					// previousTimeline.current
					// 	.totalProgress(1) // jump to the end of the animation (no tweening, no timing, just get there)
					// 	.kill(); // stop animating
				}
				previousTimeline.current = timeline;

				const nextTLs = new Map(previousTLs.current);
				unmovedCards.forEach(({ shorthand, top, left, zIndex }) => {
					// XXX (animation) should this be in animUpdatedCardPositions somehow?
					const cardId = '#' + calcCardId(shorthand, gameBoardIdRef?.current);
					timeline.set(cardId, { top, left, zIndex });
				});
				if (updateCardPositionsPrev) {
					// XXX (techdebt) (motivation) this needs to be refactored this is the first non-trivial animation, so it's a bit of a 1-off
					//  - everything else so far has been about making sure the cards move in the right order
					timeline.addLabel('updateCardPositionsPrev');
					animUpdatedCardPositions({
						timeline,
						list: updateCardPositionsPrev,
						nextTLs,
						fixtureSizes,
						prevFixtureSizes,
						gameBoardIdRef,
					});
				}
				timeline.addLabel('updateCardPositions');
				animUpdatedCardPositions({
					timeline,
					list: updateCardPositions,
					nextTLs,
					fixtureSizes,
					prevFixtureSizes,
					gameBoardIdRef,
					pause: secondMustComeAfter,
				});
				previousTLs.current = nextTLs;
			}

			if (invalidMoveCards?.fromShorthands.length) {
				if (previousTimeline.current && previousTimeline.current !== timeline) {
					// FIXME test
					previousTimeline.current.timeScale(MULTI_ANIMATION_TIMESCALE);
					// previousTimeline.current
					// 	.totalProgress(1) // jump to the end of the animation (no tweening, no timing, just get there)
					// 	.kill(); // stop animating
				}
				previousTimeline.current = timeline;

				timeline.addLabel('invalidMoveCards.fromShorthands');
				animShakeCard({
					timeline,
					list: invalidMoveCards.fromShorthands,
					gameBoardIdRef,
				});
				if (invalidMoveCards.toShorthands.length) {
					timeline.addLabel('invalidMoveCards.toShorthands');
					animShakeCard({
						timeline,
						list: invalidMoveCards.toShorthands,
						gameBoardIdRef,
					});
				}
			}
		},
		{ dependencies: [cards, selection, previousAction, fixtureSizes] }
	);

	return { previousTLs };
}
