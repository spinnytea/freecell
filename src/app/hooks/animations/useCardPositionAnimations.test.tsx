import { render } from '@testing-library/react';
import { gsap } from 'gsap/all';
import { ACTION_TEXT_EXAMPLES } from '@/app/components/cards/constants_test';
import { FreeCell } from '@/app/game/game';
import { useCardPositionAnimations } from '@/app/hooks/animations/useCardPositionAnimations';
import { FixtureSizesContextProvider } from '@/app/hooks/contexts/FixtureSizes/FixtureSizesContextProvider';
import StaticGameContextProvider from '@/app/hooks/contexts/Game/StaticGameContextProvider';

jest.mock('gsap/all', () => ({
	gsap: {
		timeline: () => ({}),
	},
}));

function MockGamePage({
	gameStateOne,
	gameStateTwo,
}: {
	gameStateOne: FreeCell | string;
	gameStateTwo: FreeCell | string;
}) {
	return (
		<StaticGameContextProvider games={[gameStateOne, gameStateTwo]}>
			<FixtureSizesContextProvider gameBoardRef={{ current: null }} fixtureLayout="portrait">
				<MockGameBoard />
			</FixtureSizesContextProvider>
		</StaticGameContextProvider>
	);
}

function MockGameBoard() {
	useCardPositionAnimations();
	return null;
}

describe('useCardPositionAnimations', () => {
	// we can use this for a lot of tests, so we don't need to keep remaking it
	const newGameState = new FreeCell();

	let fromToSpy: jest.SpyInstance;
	let toSpy: jest.SpyInstance;
	let setSpy: jest.SpyInstance;
	beforeEach(() => {
		fromToSpy = jest.fn();
		toSpy = jest.fn();
		setSpy = jest.fn();
		jest.spyOn(gsap, 'timeline').mockImplementation(() => {
			const timelineMock: unknown = {
				fromTo: fromToSpy,
				to: toSpy,
				set: setSpy,
				totalProgress: () => ({
					kill: () => {
						/* empty */
					},
				}),
			};
			return timelineMock as gsap.core.Timeline;
		});
	});

	function mockReset() {
		toSpy.mockReset();
		fromToSpy.mockReset();
		setSpy.mockReset();
	}

	function getCardIdsFromSpy(spy: jest.SpyInstance) {
		return spy.mock.calls.map(([cardId]: [string]) => cardId);
	}

	test.todo('change fixtureSizes');

	// REVIEW (animation) all of (each PreviousActionType) should be on manualtesting/page
	describe('game (each PreviousActionType)', () => {
		describe('init', () => {
			test('undefined -> init', () => {
				const gameStateOne = newGameState;
				const gameStateTwo = gameStateOne;
				expect(gameStateTwo.previousAction).toEqual({
					text: 'init',
					type: 'init',
				});

				const { rerender } = render(
					<MockGamePage gameStateOne={gameStateOne} gameStateTwo={gameStateTwo} />
				);

				expect(toSpy).not.toHaveBeenCalled();
				expect(fromToSpy).not.toHaveBeenCalled();
				expect(setSpy.mock.calls.length).toBe(52);
				expect(setSpy.mock.calls).toMatchSnapshot();
				// expect(setSpy.mock.calls.map(([cardId]) => cardId as string)).toEqual(['#cAC', '#cAD', ..., '#cKH', '#cKS']);

				mockReset();
				rerender(<MockGamePage gameStateOne={gameStateOne} gameStateTwo={gameStateTwo} />);

				expect(toSpy).not.toHaveBeenCalled();
				expect(fromToSpy).not.toHaveBeenCalled();
				expect(setSpy).not.toHaveBeenCalled();
			});

			test('win -> init', () => {
				const gameStateOne =
					'>            KC KD KH KS \n' + //
					'                         \n' + //
					':    Y O U   W I N !    :\n' + //
					'                         \n' + //
					' hand-jammed';
				const gameStateTwo = newGameState;
				expect(gameStateTwo.previousAction).toEqual({
					text: 'init',
					type: 'init',
				});

				const { rerender } = render(
					<MockGamePage gameStateOne={gameStateOne} gameStateTwo={gameStateTwo} />
				);
				mockReset();
				rerender(<MockGamePage gameStateOne={gameStateOne} gameStateTwo={gameStateTwo} />);

				// TODO (animation) (techdebt) isn't this backwards?
				//  - shouldn't kings be first?
				expect(toSpy.mock.calls.length).toBe(52);
				expect(toSpy.mock.calls[0]).toEqual([
					'#cAC',
					{ duration: 0.15, ease: 'none', zIndex: 0 },
					'<',
				]);
				expect(fromToSpy.mock.calls.length).toBe(52);
				expect(fromToSpy.mock.calls[0]).toEqual([
					'#cAC',
					{ top: 24.4, left: 428.07 },
					{ top: 453.6, left: 14.035, duration: 0.3, ease: 'power1.out' },
					'>0',
				]);
				expect(fromToSpy.mock.calls[1]).toEqual([
					'#cAD',
					{ top: 24.4, left: 519.298 },
					{ top: 453.6, left: 14.035, duration: 0.3, ease: 'power1.out' },
					'<0.006',
				]);
				expect(fromToSpy.mock.calls[51]).toEqual([
					'#cKS',
					{ top: 24.4, left: 701.754 },
					{ top: 453.6, left: 14.035, duration: 0.3, ease: 'power1.out' },
					'<0.006',
				]);
				expect(setSpy).not.toHaveBeenCalled();
			});

			// we (read: _i_) very often win a game then set it aside
			// after coming back later, maybe react has dropped the previous TLs
			//
			test.todo('win -> undefined init');
		});

		test('shuffle', () => {
			const gameStateOne = newGameState;
			const gameStateTwo = gameStateOne.shuffle32(5);
			expect(gameStateTwo.previousAction).toEqual({
				text: 'shuffle deck (5)',
				type: 'shuffle',
			});

			const { rerender } = render(
				<MockGamePage gameStateOne={gameStateOne} gameStateTwo={gameStateTwo} />
			);
			mockReset();
			rerender(<MockGamePage gameStateOne={gameStateOne} gameStateTwo={gameStateTwo} />);

			expect(toSpy).not.toHaveBeenCalled();
			expect(fromToSpy).not.toHaveBeenCalled();
			expect(setSpy).not.toHaveBeenCalled();
		});

		test('deal', () => {
			const gameStateOne = newGameState;
			const gameStateTwo = gameStateOne.dealAll();
			expect(gameStateTwo.previousAction).toEqual({
				text: 'deal all cards',
				type: 'deal',
			});

			const { rerender } = render(
				<MockGamePage gameStateOne={gameStateOne} gameStateTwo={gameStateTwo} />
			);
			mockReset();
			rerender(<MockGamePage gameStateOne={gameStateOne} gameStateTwo={gameStateTwo} />);

			expect(toSpy.mock.calls.length).toBe(52);
			expect(toSpy.mock.calls[0]).toEqual([
				'#cQC',
				{ duration: 0.15, ease: 'none', zIndex: 0 },
				'<',
			]);
			expect(fromToSpy.mock.calls.length).toBe(52);
			expect(fromToSpy.mock.calls[0]).toEqual([
				'#cQC',
				{ top: 453.6, left: 14.035 },
				{ top: 183, left: 701.754, duration: 0.3, ease: 'power1.out' },
				'>0',
			]);
			expect(fromToSpy.mock.calls[1]).toEqual([
				'#cQD',
				{ top: 453.6, left: 14.035 },
				{ top: 183, left: 603.509, duration: 0.3, ease: 'power1.out' },
				'<0.006',
			]);
			expect(fromToSpy.mock.calls[51]).toEqual([
				'#cAS',
				{ top: 453.6, left: 14.035 },
				{ top: 329.4, left: 14.035, duration: 0.3, ease: 'power1.out' },
				'<0.006',
			]);
			expect(setSpy).not.toHaveBeenCalled();
		});

		test('cursor', () => {
			const gameStateOne = newGameState.dealAll().moveCursor('down');
			const gameStateTwo = gameStateOne.moveCursor('right');
			expect(gameStateTwo.previousAction).toEqual({
				text: 'cursor right',
				type: 'cursor',
			});

			const { rerender } = render(
				<MockGamePage gameStateOne={gameStateOne} gameStateTwo={gameStateTwo} />
			);
			mockReset();
			rerender(<MockGamePage gameStateOne={gameStateOne} gameStateTwo={gameStateTwo} />);

			expect(toSpy).not.toHaveBeenCalled();
			expect(fromToSpy).not.toHaveBeenCalled();
			expect(setSpy).not.toHaveBeenCalled();
		});

		test.todo('select');

		test.todo('deselect');

		test.todo('move');

		describe('move-foundation', () => {
			test('one and one', () => {
				const gameStateOne = new FreeCell().shuffle32(5).dealAll();
				const gameStateTwo = gameStateOne.moveByShorthand('53');
				expect(gameStateTwo.previousAction).toEqual({
					text: 'move 53 6H→7C (auto-foundation 2 AD)',
					type: 'move-foundation',
					actionPrev: [
						{ rank: '6', suit: 'hearts', location: { fixture: 'cascade', data: [2, 7] } },
					],
				});

				const { rerender } = render(
					<MockGamePage gameStateOne={gameStateOne} gameStateTwo={gameStateTwo} />
				);
				mockReset();
				rerender(<MockGamePage gameStateOne={gameStateOne} gameStateTwo={gameStateTwo} />);

				expect(toSpy.mock.calls).toEqual([
					['#c6H', { zIndex: 7, duration: 0.15, ease: 'none' }, '<'],
					['#cAD', { zIndex: 99, duration: 0.15, ease: 'none' }, '<'],
				]);
				expect(fromToSpy.mock.calls).toEqual([
					[
						'#c6H',
						{ top: 305, left: 407.018 },
						{ top: 353.8, left: 210.526, duration: 0.3, ease: 'power1.out' },
						'>0',
					],
					[
						'#cAD',
						{ top: 329.4, left: 112.281 },
						{ top: 24.4, left: 428.07, duration: 0.3, ease: 'power1.out' },
						'>0',
					],
				]);
				const setCardIds = getCardIdsFromSpy(setSpy);
				expect(setCardIds.length).toBe(50);
				expect(setCardIds).not.toContain('#cAD'); // 51
				expect(setCardIds).not.toContain('#c6H'); // 52
			});

			test('multiple with overlap', () => {
				const gameStateOne = FreeCell.parse(
					'' + //
						'             QC TD KH TS \n' + //
						' KD KC       KS          \n' + //
						' QS>QD|                  \n' + //
						' JD|JS|                  \n' + //
						' select 2 QD-JS'
				);
				const gameStateTwo = gameStateOne.autoMove();
				expect(gameStateTwo.print()).toBe(
					'' +
						'             KC KD KH KS \n' + //
						'            >            \n' + //
						':    Y O U   W I N !    :\n' + //
						'                         \n' + //
						' move 25 QD-JS→KS (auto-foundation 1551215 JD,JS,QD,QS,KC,KD,KS)'
				);
				expect(gameStateTwo.previousAction).toEqual({
					text: 'move 25 QD-JS→KS (auto-foundation 1551215 JD,JS,QD,QS,KC,KD,KS)',
					type: 'move-foundation',
					actionPrev: [
						{ rank: 'queen', suit: 'diamonds', location: { fixture: 'cascade', data: [4, 1] } },
						{ rank: 'jack', suit: 'spades', location: { fixture: 'cascade', data: [4, 2] } },
					],
				});

				const { rerender } = render(
					<MockGamePage gameStateOne={gameStateOne} gameStateTwo={gameStateTwo} />
				);
				mockReset();
				rerender(<MockGamePage gameStateOne={gameStateOne} gameStateTwo={gameStateTwo} />);

				expect(toSpy.mock.calls.length).toBe(16);
				expect(toSpy.mock.calls).toEqual([
					['#cQD', { zIndex: 1, duration: 0.15, ease: 'none' }, '<'],
					['#cJS', { zIndex: 2, duration: 0.15, ease: 'none' }, '<'],
					// and then
					['#cJD', { top: 24.4, left: 519.298, duration: 0.3, ease: 'power1.out' }, '>0.043'],
					['#cJD', { zIndex: 109, duration: 0.15, ease: 'none' }, '<'],
					['#cJS', { top: 24.4, left: 701.754, duration: 0.3, ease: 'power1.out' }, '<0.043'],
					['#cJS', { zIndex: 109, duration: 0.15, ease: 'none' }, '<'],
					['#cQD', { top: 24.4, left: 519.298, duration: 0.3, ease: 'power1.out' }, '<0.043'],
					['#cQD', { zIndex: 110, duration: 0.15, ease: 'none' }, '<'],
					['#cQS', { top: 24.4, left: 701.754, duration: 0.3, ease: 'power1.out' }, '<0.043'],
					['#cQS', { zIndex: 110, duration: 0.15, ease: 'none' }, '<'],
					['#cKC', { top: 24.4, left: 428.07, duration: 0.3, ease: 'power1.out' }, '<0.043'],
					['#cKC', { zIndex: 111, duration: 0.15, ease: 'none' }, '<'],
					['#cKD', { top: 24.4, left: 519.298, duration: 0.3, ease: 'power1.out' }, '<0.043'],
					['#cKD', { zIndex: 111, duration: 0.15, ease: 'none' }, '<'],
					['#cKS', { top: 24.4, left: 701.754, duration: 0.3, ease: 'power1.out' }, '<0.043'],
					['#cKS', { zIndex: 111, duration: 0.15, ease: 'none' }, '<'],
				]);
				expect(fromToSpy.mock.calls).toEqual([
					[
						'#cQD',
						{ top: 201.3, left: 112.281 },
						{ top: 207.4, left: 407.018, duration: 0.3, ease: 'power1.out' },
						'>0',
					],
					[
						'#cJS',
						{ top: 244, left: 112.281 },
						{ top: 231.8, left: 407.018, duration: 0.3, ease: 'power1.out' },
						'<0.060',
					],
				]);
				const setCardIds = getCardIdsFromSpy(setSpy);
				expect(setCardIds.length).toBe(45);
				expect(setCardIds).not.toContain('#cJD'); // 46
				expect(setCardIds).not.toContain('#cJS');
				expect(setCardIds).not.toContain('#cQD');
				expect(setCardIds).not.toContain('#cQS');
				expect(setCardIds).not.toContain('#cKC'); // 50
				expect(setCardIds).not.toContain('#cKD');
				expect(setCardIds).not.toContain('#cKS'); // 52
			});
		});

		test.todo('move-flourish');

		test.todo('auto-foundation');

		test.todo('auto-flourish');

		describe('invalid (each MoveSourceType)', () => {
			test.todo('deck');

			test.todo('cell');

			test.todo('foundation');

			test.todo('cascade:single');

			test.todo('cascade:sequence');
		});
	});

	describe('actionText examples', () => {
		// let actionTextExamples: string[];
		// beforeAll(() => {
		// 	actionTextExamples = ACTION_TEXT_EXAMPLES.slice(0);
		// });
		// afterAll(() => {
		// 	expect(actionTextExamples).toEqual([]);
		// });

		// TODO (techdebt) (animation) actually write the tests for these?
		ACTION_TEXT_EXAMPLES.forEach((actionText) => {
			test.todo(actionText + '');
		});
	});

	// manual testing scenarios are for visual fine-tuning
	// these may be [basically] duplicates of other unit tests
	describe('manual testing scenarios', () => {
		test.todo('52CardFlourish');

		test.todo('readyToAutoFoundation');

		test.todo('win foundation -> deck');

		test.todo('animations 0');

		test.todo('animations 1');

		test.todo('animations 2');
	});

	describe('bugfix', () => {
		test('Invalid Undo 2S (animation)', () => {
			const gameStateOne = FreeCell.parse(
				'' +
					'    3H 8D 4D AC 2D AH 2S \n' +
					'    JC 9D 9C KD KC KS 5C \n' +
					'    JD 8S 4C QS    QH 2H \n' +
					'    6D 7D 3C       JS TD \n' +
					'    6H 6C 7S       TH 9S \n' +
					'    QC 5H QD          2C \n' +
					'    KH    TS          5S \n' +
					'    8H    JH          4H \n' +
					'    7C    TC          3S \n' +
					'          9H             \n' +
					'          8C             \n' +
					'          7H             \n' +
					'          6S             \n' +
					'          5D             \n' +
					'          4S             \n' +
					'          3D             \n' +
					' move 14 2S→3D (auto-foundation 14 AS,2S)\n' +
					':h shuffle32 2107\n' +
					' 64 62 6a 6b 3c 34 14 74 \n' +
					' 34 38 3d 34 18 15 73 71 \n' +
					' 73 57 53 57 54 13 a5 16 \n' +
					' 14 '
			).moveByShorthand('21');
			expect(gameStateOne.previousAction).toEqual({
				text: 'move 21 8H-7C→cascade',
				type: 'move',
			});
			const gameStateTwo = gameStateOne.undo();
			expect(gameStateTwo.previousAction).toEqual({
				text: 'move 14 2S→3D (auto-foundation 14 AS,2S)',
				type: 'move-foundation',
				actionPrev: [
					{ rank: '2', suit: 'spades', location: { fixture: 'cascade', data: [3, 15] } },
				],
			});

			// "2 of spades" does not move between states
			const gameStateOne_2S = gameStateOne.cards.find(
				({ rank, suit }) => rank === '2' && suit === 'spades'
			);
			const gameStateTwo_2S = gameStateTwo.cards.find(
				({ rank, suit }) => rank === '2' && suit === 'spades'
			);
			expect(gameStateOne_2S?.location).toEqual({ fixture: 'foundation', data: [3] });
			expect(gameStateTwo_2S?.location).toEqual({ fixture: 'foundation', data: [3] });
			expect(gameStateOne_2S?.location).toBe(gameStateOne_2S?.location);

			const { rerender } = render(
				<MockGamePage gameStateOne={gameStateOne} gameStateTwo={gameStateTwo} />
			);
			mockReset();
			rerender(<MockGamePage gameStateOne={gameStateOne} gameStateTwo={gameStateTwo} />);

			expect(toSpy.mock.calls).toEqual([
				// ['#c2S', { zIndex: 15, duration: 0.15, ease: 'none' }, '<'],
				['#c7C', { zIndex: 7, duration: 0.15, ease: 'none' }, '<'],
				['#c8H', { zIndex: 6, duration: 0.15, ease: 'none' }, '<'],
			]);
			expect(fromToSpy.mock.calls).toEqual([
				// [
				// 	'#c2S',
				// 	{ top: 24.4, left: 701.754 },
				// 	{ top: 549, left: 308.772, duration: 0.3, ease: 'power1.out' },
				// 	'>0',
				// ],
				[
					'#c7C',
					{ top: 207.4, left: 14.035 },
					{ top: 353.8, left: 112.281, duration: 0.3, ease: 'power1.out' },
					'>0',
				],
				[
					'#c8H',
					{ top: 183, left: 14.035 },
					{ top: 329.4, left: 112.281, duration: 0.3, ease: 'power1.out' },
					'<0.060',
				],
			]);
			const setCardIds = getCardIdsFromSpy(setSpy);
			expect(setCardIds.length).toBe(50);
			expect(setCardIds).not.toContain('#c7C'); // 51
			expect(setCardIds).not.toContain('#c8H'); // 52
		});
	});
});
