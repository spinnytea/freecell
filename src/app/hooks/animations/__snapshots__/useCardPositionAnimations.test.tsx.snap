// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useCardPositionAnimations game (each PreviousActionType) init undefined -> init 1`] = `
[
  [
    "#cAC",
    {
      "left": 14.035,
      "top": 441.4,
      "zIndex": 0,
    },
  ],
  [
    "#cAD",
    {
      "left": 14.035,
      "top": 441.4,
      "zIndex": 1,
    },
  ],
  [
    "#cAH",
    {
      "left": 14.035,
      "top": 441.4,
      "zIndex": 2,
    },
  ],
  [
    "#cAS",
    {
      "left": 14.035,
      "top": 441.4,
      "zIndex": 3,
    },
  ],
  [
    "#c2C",
    {
      "left": 14.035,
      "top": 441.4,
      "zIndex": 4,
    },
  ],
  [
    "#c2D",
    {
      "left": 14.035,
      "top": 441.4,
      "zIndex": 5,
    },
  ],
  [
    "#c2H",
    {
      "left": 14.035,
      "top": 441.4,
      "zIndex": 6,
    },
  ],
  [
    "#c2S",
    {
      "left": 14.035,
      "top": 441.4,
      "zIndex": 7,
    },
  ],
  [
    "#c3C",
    {
      "left": 14.035,
      "top": 441.4,
      "zIndex": 8,
    },
  ],
  [
    "#c3D",
    {
      "left": 14.035,
      "top": 441.4,
      "zIndex": 9,
    },
  ],
  [
    "#c3H",
    {
      "left": 14.035,
      "top": 441.4,
      "zIndex": 10,
    },
  ],
  [
    "#c3S",
    {
      "left": 14.035,
      "top": 441.4,
      "zIndex": 11,
    },
  ],
  [
    "#c4C",
    {
      "left": 14.035,
      "top": 441.4,
      "zIndex": 12,
    },
  ],
  [
    "#c4D",
    {
      "left": 14.035,
      "top": 441.4,
      "zIndex": 13,
    },
  ],
  [
    "#c4H",
    {
      "left": 14.035,
      "top": 441.4,
      "zIndex": 14,
    },
  ],
  [
    "#c4S",
    {
      "left": 14.035,
      "top": 441.4,
      "zIndex": 15,
    },
  ],
  [
    "#c5C",
    {
      "left": 14.035,
      "top": 441.4,
      "zIndex": 16,
    },
  ],
  [
    "#c5D",
    {
      "left": 14.035,
      "top": 441.4,
      "zIndex": 17,
    },
  ],
  [
    "#c5H",
    {
      "left": 14.035,
      "top": 441.4,
      "zIndex": 18,
    },
  ],
  [
    "#c5S",
    {
      "left": 14.035,
      "top": 441.4,
      "zIndex": 19,
    },
  ],
  [
    "#c6C",
    {
      "left": 14.035,
      "top": 441.4,
      "zIndex": 20,
    },
  ],
  [
    "#c6D",
    {
      "left": 14.035,
      "top": 441.4,
      "zIndex": 21,
    },
  ],
  [
    "#c6H",
    {
      "left": 14.035,
      "top": 441.4,
      "zIndex": 22,
    },
  ],
  [
    "#c6S",
    {
      "left": 14.035,
      "top": 441.4,
      "zIndex": 23,
    },
  ],
  [
    "#c7C",
    {
      "left": 14.035,
      "top": 441.4,
      "zIndex": 24,
    },
  ],
  [
    "#c7D",
    {
      "left": 14.035,
      "top": 441.4,
      "zIndex": 25,
    },
  ],
  [
    "#c7H",
    {
      "left": 14.035,
      "top": 441.4,
      "zIndex": 26,
    },
  ],
  [
    "#c7S",
    {
      "left": 14.035,
      "top": 441.4,
      "zIndex": 27,
    },
  ],
  [
    "#c8C",
    {
      "left": 14.035,
      "top": 441.4,
      "zIndex": 28,
    },
  ],
  [
    "#c8D",
    {
      "left": 14.035,
      "top": 441.4,
      "zIndex": 29,
    },
  ],
  [
    "#c8H",
    {
      "left": 14.035,
      "top": 441.4,
      "zIndex": 30,
    },
  ],
  [
    "#c8S",
    {
      "left": 14.035,
      "top": 441.4,
      "zIndex": 31,
    },
  ],
  [
    "#c9C",
    {
      "left": 14.035,
      "top": 441.4,
      "zIndex": 32,
    },
  ],
  [
    "#c9D",
    {
      "left": 14.035,
      "top": 441.4,
      "zIndex": 33,
    },
  ],
  [
    "#c9H",
    {
      "left": 14.035,
      "top": 441.4,
      "zIndex": 34,
    },
  ],
  [
    "#c9S",
    {
      "left": 14.035,
      "top": 441.4,
      "zIndex": 35,
    },
  ],
  [
    "#cTC",
    {
      "left": 14.035,
      "top": 441.4,
      "zIndex": 36,
    },
  ],
  [
    "#cTD",
    {
      "left": 14.035,
      "top": 441.4,
      "zIndex": 37,
    },
  ],
  [
    "#cTH",
    {
      "left": 14.035,
      "top": 441.4,
      "zIndex": 38,
    },
  ],
  [
    "#cTS",
    {
      "left": 14.035,
      "top": 441.4,
      "zIndex": 39,
    },
  ],
  [
    "#cJC",
    {
      "left": 14.035,
      "top": 441.4,
      "zIndex": 40,
    },
  ],
  [
    "#cJD",
    {
      "left": 14.035,
      "top": 441.4,
      "zIndex": 41,
    },
  ],
  [
    "#cJH",
    {
      "left": 14.035,
      "top": 441.4,
      "zIndex": 42,
    },
  ],
  [
    "#cJS",
    {
      "left": 14.035,
      "top": 441.4,
      "zIndex": 43,
    },
  ],
  [
    "#cQC",
    {
      "left": 14.035,
      "top": 441.4,
      "zIndex": 44,
    },
  ],
  [
    "#cQD",
    {
      "left": 14.035,
      "top": 441.4,
      "zIndex": 45,
    },
  ],
  [
    "#cQH",
    {
      "left": 14.035,
      "top": 441.4,
      "zIndex": 46,
    },
  ],
  [
    "#cQS",
    {
      "left": 14.035,
      "top": 441.4,
      "zIndex": 47,
    },
  ],
  [
    "#cKC",
    {
      "left": 14.035,
      "top": 441.4,
      "zIndex": 48,
    },
  ],
  [
    "#cKD",
    {
      "left": 14.035,
      "top": 441.4,
      "zIndex": 49,
    },
  ],
  [
    "#cKH",
    {
      "left": 14.035,
      "top": 441.4,
      "zIndex": 50,
    },
  ],
  [
    "#cKS",
    {
      "left": 14.035,
      "top": 441.4,
      "zIndex": 51,
    },
  ],
]
`;
